WIP: Makefile: Initialise Makefile

iammarco11 opened the pull request:
fixes #1 and #7.

Making it WIP because I don't know if it is the perfect way to do things :P

Aniketh01 commented on Makefile:
```diff
@@ -0,0 +1,30 @@
+LIBS := -lstdc++ -lavcodec -lavformat -lavutil -lswresample -lswscale -lglfw -lGL
+TARGET_DIR := bin/
+
+SRC_FILES_PLAYER := main_window.cpp decoder.cpp
+SRC_FILES_TCP_PLAYER := tcp_main_window.cpp decoder.cpp tcp_client.c
+SRC_FILES_TCP_SERVER := tcp_server.c
+SRC_FILES_TRANSCODER := transcoder.c debug.c
```
FILES is not necessary I assume?

iammarco11 replied:
aah just my bad variable naming :P

Aniketh01 replied:
No worries :)

Aniketh01 commented on Makefile:
```diff
@@ -0,0 +1,30 @@
+
+all: dir player tcp_player tcp_server transcoder
+
+dir:
+	if [ ! -d "$(TARGET_DIR)" ]; then \
+		mkdir -p $(TARGET_DIR); \
+	fi
+	
```
Can't this be automated? Rather than asking the user to do `make dir` can't we have it prepared when any of the binary is compiled?

iammarco11 replied:
the user just needs to put make and everything else will be done

Aniketh01 replied:
okay, but consider a case if the user only wants to prepare the TCP player binary and there isn't a bin/ dir. what would be the result then? Would makefile report an error?

iammarco11 replied:
because I have specified the dir target inside all

Aniketh01 replied:
You sure about that? would you mind rechecking the scenario I explained and get back to me?

iammarco11 replied:
ohhhh I get you
will work on that soon

Aniketh01 commented on Makefile:
```diff
@@ -0,0 +1,30 @@
+	fi
+	
+player:
+	$(CC) $(SRC_FILES_PLAYER) $(LIBS) -o $(TARGET_DIR)$@
```
Can you check if -o itself is the way to mention in the makefile that the output binary should land in the target dir?

Aniketh01 commented:
Looks a good start. No worries if you don't know the proper way. We only learn by doing and most importantly, by making mistakes. That is totally acceptable. But there should be progress at the end of the day :)

iammarco11 committed:
Makefile: Initialise Makefile
This commit does
- Initialise makefile and uses targets to be made
- Separate variables for source files, directories and common compiler

iammarco11 committed:
Makefile: Enable peek variable value
This commit does:
- When make print-VARIABLE
- It returns the VARIABLE = the_value_of_the_variable

Aniketh01 commented on Makefile:
```diff
@@ -0,0 +1,35 @@
+
+all: dir player tcp_player tcp_server transcoder
+
+dir:
+	if [ ! -d "$(TARGET_DIR)" ]; then \
+		mkdir -p $(TARGET_DIR); \
+	fi
+
```
Currently looks like a patchy work while redundantly using dir variable everywhere. Can we by default make dir the bin folder is created every time the make command is run no matter what binary we want to prepare?

iammarco11 replied:
I am not getting it, do you know how to? I read in stackoverflow, thats how I did the patchy work.

Aniketh01 replied:
The check you do to see if that particular directory is there or not and to create the directory, can't it be written such a way that those particular set of code executes even if 'make dir' is not provided?

iammarco11 replied:
> can't it be written such a way that those particular set of code executes even if 'make dir' is not provided?

Well the the dir is made even if we give `make player` and not `make dir` everytime because I put dir as a prerequisite to every other target file in all the rules. I agree it is patchy and redundant putting dir everywhere as a prerequisite, but I haven't come across a better solution. I will need to dig more on that then, but if you know please let me know :)

iammarco11 committed:
Makefile: Add the shell command for output directory

iammarco11 committed:
Makefile: Add clean recipe
This commit does
- Adds a rule clean which will remove the output directory

Aniketh01 commented on Makefile:
```diff
@@ -0,0 +1,35 @@
+SRC_TCP_SERVER := tcp_server.c
+SRC_TRANSCODER := transcoder.c debug.c
+
+$(shell mkdir -p $(TARGET_DIR))
+
+.PHONY: all
+
+all: player tcp_player tcp_server transcoder
```
Better than this would be to check if the target_dir is present or not in `all`. If it is not present create one and move forward. If it is present move forward.

iammarco11 replied:
Well I quite didn't understand you, did you mean putting target_dir as a pre requisite for `all` and add the `mkdir` as a recipe for the `all`. If that's what you meant, `make player` won't create the directory.
